psycopg/psycopg/connection.py

```diff
@@ -95,27 +95,44 @@ def connect(
         timeout = timeout_from_conninfo(params)
         rv = None
         attempts = conninfo_attempts(params)
+        connection_errors: list[tuple[e._NoTracebackException, str]] = []
         for attempt in attempts:
             try:
                 conninfo = make_conninfo("", **attempt)
                 gen = cls._connect_gen(conninfo, timeout=timeout)
                 rv = waiting.wait_conn(gen, interval=_WAIT_INTERVAL)
             except e._NO_TRACEBACK as ex:
-                if len(attempts) > 1:
-                    logger.debug(
-                        "connection attempt failed: host: %r port: %r, hostaddr %r: %s",
-                        attempt.get("host"),
-                        attempt.get("port"),
-                        attempt.get("hostaddr"),
-                        str(ex),
-                    )
-                last_ex = ex
+                attempt_details = "host: {}, port: {}, hostaddr: {}".format(
+                    repr(attempt.get("host")),
+                    repr(attempt.get("port")),
+                    repr(attempt.get("hostaddr")),
+                )
+                connection_errors.append((ex, attempt_details))
             else:
                 break
 
         if not rv:
-            assert last_ex
-            raise last_ex.with_traceback(None)
+            num_errors = len(connection_errors)
+            assert num_errors > 0
```

Comment:
I don't think this assert is useful. `assert last_ex` was necessary to please MyPy, but this list will never be empty, and even if it was, IndexError or AssertError don't make any difference.

Reply:
Agreed, I can remove this.

Reply:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)